server-utils/src/suspendable_stream.rs

```diff
@@ -47,7 +47,10 @@ impl<S, I> Stream for SuspendableStream<S>
 					self.timeout = Some(timeout);
 					return Ok(Async::NotReady);
 				}
-				Err(_) => unreachable!("Polling a delay shouldn't yield any errors; qed")
+				Err(_) => {
+					// we've reached maximum number of concurrent timers; drop the timeout,
```

Comment:
That is not good, as the `Stream` will not be called anymore by the tokio runtime. What we could do is the following:
```
// Wake your `Stream` is the next round again (poll is called again)
futures::task::current().notify();
```
Then we also add some sort of counter, that counts how many successive calls we have, without being able to register a delay or polling the underlying socket. After around 1000 cycles, we should try something different, like maybe putting the whole thread sleeping and hoping that some other thread is able to finish its work and we are able to accept new connections.

Reply:
> That is not good, as the `Stream` will not be called anymore by the tokio runtime. 

can you please explain how?

Reply:
You mean why the `Stream` will not be called anymore by Tokio?

Tokio uses internally `mio` and that uses `epoll` on Linux. Epoll will just notify you, if you have polled the resource and you were instructed to wait. If you polled the resource and got an error, you are probably not registered anymore to be informed about any new events that happen.

Reply:
> if you polled the resource and got an error, you are probably not registered anymore

Streams don't terminate when they encounter an error. 
https://docs.rs/futures/0.1.21/futures/stream/trait.Stream.html#errors

and the socket actually has nothing to do with with the timeout, I think if you read the full code you'd have a better understanding

https://github.com/paritytech/jsonrpc/blob/f42e5e65f7f2723879f784505a9151d87bcd1397/server-utils/src/suspendable_stream.rs#L42-L81

If we attempt to poll the timeout and it yields an error (which in this case would be an `at_capacity` error). there's nothing we can do other than to shed load, by dropping the timeout and proceeding to resume polling the stream.

Reply:
You are right, the `futures` documentation says that `Stream` does not terminate when you get an error. But, I talked about `Tokio` and that is just an implementation of a `futures::Executor`. Let's take an example, let's say you want to wrap a `Stream` in a `Future`. The `Future` will just `poll` the `Stream` and should return `Ok(Ready())` when the `Stream is finished, but you want also make sure that you at most pull out 5 items from the `Stream` at a time. So you would do something like:
```
fn poll(&mut self) -> Poll<> {
    for _ in 0..5 {
         let item = try_ready!(self.stream.poll());
         if item.is_none() { return Ok(Ready(()); }
         self.handle_item(item);
    }

    return Ok(NotReady());
}
````

In this example, you can end in a situation where your future is not polled anymore. This will happen, when you pull out 5 items in a row, but the `Stream` has still more items to be pulled. But you are not telling the `Tokio` runtime that your future should be called again and thus, you will not get called again. 
I know that is really complicated and I already suffered several times from random bugs (futures where not called anymore etc), because of all these implementation details you need to know.

Reply:
I also looked again over your code and when you enter the state where you create a timeout, you will just loop forever.
You need to leave the loop, after you created your timeout and you also need to poll your timeout at least once, that it is registered in the runtime and your futures gets called when the timeout is reached.

Reply:
> and you also need to poll your timeout at least once

yeah i completely forgot about it, @tomusdrw asked to remove the duplicate polling logic and it skipped my mind to wrap the whole thing inside the loop